cmd/update/volume_move.go

```diff
@@ -0,0 +1,383 @@
+	"code.storageos.net/storageos/c2-cli/pkg/health"
+	"code.storageos.net/storageos/c2-cli/pkg/id"
+	"code.storageos.net/storageos/c2-cli/pkg/version"
+)
+
+type volumeMoveCommand struct {
+	config  interfaces.ConfigProvider
+	client  interfaces.Client
+	display interfaces.Displayer
+
+	// inputs
+	namespace string
+	volumeID  string
+	timeout   time.Duration
+	fromNode  string
+	toNode    string
+
+	// useCAS determines whether the command makes the request constrained by
+	// the provided casVersion.
+	useCAS     func() bool
+	casVersion string
+	useAsync   bool
+
+	writer io.Writer
+}
+
+func (c *volumeMoveCommand) runWithCtx(ctx context.Context, cmd *cobra.Command, args []string) error {
+	useIDs, err := c.config.UseIDs()
+	if err != nil {
+		return err
+	}
+
+	// fetch IDs
+	nsID := id.Namespace(c.namespace)
+
+	if !useIDs {
+		ns, err := c.client.GetNamespaceByName(ctx, c.namespace)
+		if err != nil {
+			return err
+		}
+		nsID = ns.ID
+
+		fromNode, err := c.client.GetNodeByName(ctx, c.fromNode)
+		if err != nil {
+			return err
+		}
+		c.fromNode = fromNode.ID.String()
+
+		toNode, err := c.client.GetNodeByName(ctx, c.toNode)
+		if err != nil {
+			return err
+		}
+		c.toNode = toNode.ID.String()
+	}
+
+	// fetch volume
+	vol, err := c.getVolume(ctx, nsID)
+	if err != nil {
+		return err
+	}
+
+	if vol.Master.Health != health.MasterOnline {
+		return fmt.Errorf("cannot move volume - master unhealthy: %v", vol.Master.Health.String())
+	}
+	for _, replica := range vol.Replicas {
+		if replica.Health != health.ReplicaReady || !replica.Promotable {
+			return fmt.Errorf("cannot move volume - replica unhealthy, id: %v health: %v, promotable: %v", replica.ID, replica.Health, replica.Promotable)
+		}
+	}
+
+	hasReplicaOnSrcNode := false
+	isRetryRun := false
+	var deploymentOnSourceNode id.Deployment
+	for _, rep := range vol.Replicas {
+		// make sure there's at least one replica on the [source] node
+		if rep.Node.String() == c.fromNode {
+			hasReplicaOnSrcNode = true
+			deploymentOnSourceNode = rep.ID
+		}
+
+		// if there's already a replica on the [destination] node we will assume
+		// we are retrying a previously failed request (timed out would be the
+		// most likely failure)
+		if rep.Node.String() == c.toNode {
+			isRetryRun = true
+		}
+	}
+
+	isMovingMaster := false
+	if vol.Master.Node.String() == c.fromNode {
+		isMovingMaster = true
+	} else if !hasReplicaOnSrcNode {
+		return fmt.Errorf("volume has no replicas on [from] node %s", c.fromNode)
+	}
+
+	// create the new replica on [destination] node
+	// skip if there's one one already (retry run)
+	if !isRetryRun {
+		params := &apiclient.AddDeploymentOnNodeRequestParams{
+			NodeID: id.Node(c.toNode),
+		}
+
+		if c.useCAS() {
+			params.CASVersion = version.FromString(c.casVersion)
+		}
+
+		err = c.client.AddDeploymentOnNode(ctx, nsID, vol.ID, params)
+		if err != nil {
+			return fmt.Errorf("failed to add deployment on node %s: %s", c.toNode, err)
```

Comment:
nit: it's good to use `%w` instead of `%s` to wrap errors, as it allows us to use `errors.Is` and `errors.As`. It's not too relevant here, as we don't use those functions, but it's good practice

https://go.dev/blog/go1.13-errors

Reply:
Updated to use `%w`